src/Microsoft.AspNet.Razor/Generator/Compiler/CodeBuilder/CSharp/CSharpTagHelperCodeRenderer.cs

```diff
@@ -190,81 +190,126 @@ protected virtual string GenerateUniqueId()
 
         private void RenderBoundHTMLAttributes(
             IList<KeyValuePair<string, Chunk>> chunkAttributes,
+            string tagHelperTypeName,
             string tagHelperVariableName,
             IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
             Dictionary<string, string> htmlAttributeValues)
         {
-            foreach (var attributeDescriptor in attributeDescriptors)
+            // Track dictionary properties we have confirmed are non-null.
+            var confirmedDictionaries = new HashSet<string>(StringComparer.Ordinal);
+
+            // First attribute wins, even if there are duplicates.
+            var distinctAttributes = chunkAttributes.Distinct(KeyValuePairKeyComparer.Default);
+
+            // Go through the HTML attributes in source order, assigning to properties or indexers as we go.
+            foreach (var attributeKeyValuePair in distinctAttributes)
             {
-                var matchingAttributes = chunkAttributes.Where(
-                    kvp => string.Equals(kvp.Key, attributeDescriptor.Name, StringComparison.OrdinalIgnoreCase));
+                var attributeName = attributeKeyValuePair.Key;
+                var attributeValueChunk = attributeKeyValuePair.Value;
+                if (attributeValueChunk == null)
+                {
+                    // Minimized attributes are not valid for bound attributes. TagHelperBlockRewriter has already
+                    // logged an error if it was a bound attribute; so we can skip.
+                    continue;
+                }
 
-                if (matchingAttributes.Any())
+                // Find the matching TagHelperAttributeDescriptor.
+                var attributeDescriptor = attributeDescriptors.FirstOrDefault(
```

Comment:
Kinda hack since this is relying on the Indexers always being at the end of the attribute descriptors. Is the plan to eventually create an error if multiple attribute descriptors for a single `TagHelperDescriptor` target the same attribute?